annotation-processing/src/main/java/com/exactpro/epfast/annotation/processing/FastProcessor.java

```diff
@@ -19,112 +23,198 @@
-                    fastTypesByName.values().stream().toArray(Element[]::new));
+                addServiceEntries(SERVICE_TYPENAME, creatorImplTypeNames,
+                    originatingElements.stream().toArray(Element[]::new));
             } catch (IOException e) {
-                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR,
+                messager.printMessage(Diagnostic.Kind.ERROR,
                     String.format("Unable to create or update java services file for %s service", SERVICE_TYPENAME));
             }
         }
+
         if (annotations.contains(fastTypeElement)) {
-            if (fastTypesByName != null) {
-                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR,
+            if (packageElementsMap != null) {
+                messager.printMessage(Diagnostic.Kind.ERROR,
                     "@FastType annotation processor can not handle classes with @FastType " +
                         "annotation added after first processing round. " +
                         "It may be caused by activity of other annotation processors.");
             }
-            fastTypesByName = collectFastTypes(roundEnv);
+            packageElementsMap = collectFastPackages(roundEnv);
+            HashMap<String, Element> fastTypesByName = collectFastTypes(roundEnv);
+            organizeInPackages(fastTypesByName);
             try {
-                buildCreatorClass(fastTypesByName);
+                buildCreatorClass(packageElementsMap);
             } catch (Exception e) {
-                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, e.getMessage());
+                messager.printMessage(Diagnostic.Kind.ERROR, e.getMessage());
             }
             return true;
         }
-
         return false;
     }
 
+    private void organizeInPackages(HashMap<String, Element> fastTypesByName) {
+        for (Map.Entry<String, Element> entry : fastTypesByName.entrySet()) {
+            PackageElement enclosingPackage = elementUtils.getPackageOf(entry.getValue());
+            while (!packageElementsMap.containsKey(enclosingPackage)) {
+                enclosingPackage = getParent(enclosingPackage);
+            }
+            packageElementsMap.get(enclosingPackage).put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    private PackageElement getParent(Element enclosing) {
+        String currentPackage = enclosing.toString();
+        int lastDotIndex = currentPackage.lastIndexOf('.');
+        if (lastDotIndex  == -1) {
+            return elementUtils.getPackageElement("");
+        }
+        return elementUtils.getPackageElement(currentPackage.substring(0, lastDotIndex));
+    }
+
+    private HashMap<PackageElement, HashMap<String, Element>> collectFastPackages(RoundEnvironment roundEnv) {
+        HashMap<PackageElement, HashMap<String, Element>> packageElementMap = new HashMap<>();
+        PackageElement topPackage = elementUtils.getPackageElement("");
```

Comment:
topPackage for Fast types isn't necessary root java package: it's the innermost package containing all packages with FastType annotated classes which don't belong to any FastPackage annotated package, i.e. for classes a.b.TypeOne, a.c.TypeTwo, topPackage should be "a"